### ISSUE TITLE
domvm 2.x-dev (part 2)

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/domvm#101]
(continuation of #100 since Github cannot be bothered with any kind of trivial pagination)

**Comment by leeoniya** [comment]
Now that 2.x's view layer is basically feature-complete. I'm working on improving ergonomics of prop, sync, fetch, route.

In short, streams are nice. They allow domvm to do some dependency resolution and auto-redraw only those views that return streams in their render() templates. If you pass a stream directly into a view, rather than its value (which you can always get by invoking it), domvm will hook up an auto-redraw whenever the property changes value. If the stream is used across multiple views, then only those views will be redrawn on mutations.

demo: https://jsfiddle.net/LqxhL5yf/

```js
var view = domvm.view;
var el = view.defineElement;

function View(vm, pers) {
  return function() {
	// use the stream directly for auto-redraw hookup, NOT its value: pers.age()
    return el("div", pers.age);
  }
}

var pers = {
  age: flyd.stream(0)
};

var vm = view.createView(View, pers).mount(document.body);

// increment age prop
setInterval(function() {
  pers.age(pers.age()+1);
}, 250);
```

**Comment by lawrence-dol** [comment]
> This config object can be returned from the view closure instead of render. It's only done during view init, so there's no perf impact. Returning just render will still work as before.

Oh good.

> The new functionality allows for cleaner-looking inits

I really like that.

**Comment by yosbelms** [comment]
> The plan is to provide a stream support for template-bound events, attr values, css prop values, ajax responses and route changes

Great!!

**Comment by leeoniya** [comment]
> `age: flyd.stream(0)`

Looks like i may have to abstract `flyd` behind some `.prop()` to avoid future breakage and to allow swapping out the underlying streams implementation. Most stream libs seem to be coalescing around the fantasy-land spec [1] (to which flyd does not currently conform). On the other hand I don't want domvm to expose a neutered stream impl and I'm not about to create that. I know Leo implemented some subset of the spec in Mithril rewrite [2], but I'd prefer to ingest spec-compliant streams from a third-party lib that i don't have to maintain and that dont need to be created using a domvm abstraction (like domvm.prop). Ideally flyd would implement the 2.0 spec, but the project is not particularly active and the author is waiting on some other stuff to settle [3]. Not sure what a good path forward would be. :/

[1] https://github.com/fantasyland/fantasy-land
[2] https://github.com/lhorie/mithril.js/blob/rewrite/docs/prop.md
[3] https://github.com/fantasyland/fantasy-land/issues/194

**Comment by lawrence-dol** [comment]
> Looks like i may have to abstract flyd...

Is the problem deeper than just aliasing `flyd` to something DOMVM specific? That is is the problem the use of `flyd` as an identifier?

**Comment by leeoniya** [comment]
mostly that i dont want to be married to the flyd and its current api if it will be superseded by an official spec. pulling in a stream lib as a dep should bring more benefits than just domvm's limited use-case. i would rather pull in something that complies to a spec.

all that being said, i think there's an easy solution here. domvm doesnt need deep spec compliance, only the following which seem to be pretty well implemented cross-lib:

- `stream()` to get the value
- `stream(val)` to set the value
- `stream.combine()` to create dependent streams / subscribe
- `stream.end()` to terminate a stream / unsubscribe
- `isStream(val)` to identify if passed val is a stream

domvm can provide an adapter spec/interface so any stream lib can be used that provides these minimums. a flyd adapter would be created for now, but you could BYO streams by writing a trivial adapter in the future. this would allow for devs to use streams without a domvm proxy and to allow domvm to work with them as well.

**Comment by leeoniya** [comment]
So the included flyd adapter currently looks like this [1]:

```js
import { streamCfg } from './streamCfg';

if (typeof flyd != "undefined") {
	streamCfg({
		is:		s => flyd.isStream(s),
		val:	s => s(),
		sub:	(s, fn) => flyd.on(fn, s),
		unsub:	s => s.end(true),
	});
}
```

Alternate adapters can be written externally via the exposed `domvm.streamCfg()`.

- `val` (getting the stream's value) will be available in all stream impls in some form.
- `sub` (returning a chained/dependent stream) will be available in all stream impls in some form.
- `is` may not be available, but you can use some heuristics to discern a stream without too much trouble, eg: `.isFunc(s) && isFunc(s.combine)`
- `unsub` **i'm unsure about**. flyd offers methods for ending a dependent stream and severing the subscription/dependency. i was unable to find anything that would allow for doing this in fantasy-land spec, though it's a pretty unfamiliar read for me. RxJS *does* have a dispose/unsubscribe [2] as does Bacon.js [3].

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/src/streamFlyd.js
[2] https://github.com/ReactiveX/rxjs/blob/master/doc/observable.md#disposing-observable-executions
[3] https://github.com/baconjs/bacon.js#cleaning-up

**Comment by leeoniya** [comment]
done:

- streams as full body: `el("div", stream)` or `tx(stream)`
- streams as attr: `el("input", {disabled: stream})`
- streams as css prop: `el("input", {style: {background: stream}})`

demo: https://github.com/leeoniya/domvm/blob/2.x-dev/demos/streams.html

todo:

- streams as dom event handlers, sync
- streams as route callbacks
- streams as fetch callbacks

maybe?

- streams as body nodes: `el("div", [stream, stream])`

**Comment by yosbelms** [comment]
What I see is that domvm is getting coupled to streams. This coupling could impact negatively in the code maintenance, because obvious reasons. So, what about reacting on streams changes?
```js
function View(vm) {
    vm.reactOnChange(stream1, stream2)
    // or vm.listen ...

    return function() {
        return el('input', {disabled: stream1}, stream2)
    }
}
```

Following that line domvm has to check just if the value is a function:

```js
if (isFunction(value)) {
    value = value()
}

//use `value`
```

At this point is also coupled, but with a single coupling point. Any thought on this?

**Comment by leeoniya** [comment]
@yosbelms 

> What I see is that domvm is getting coupled to streams.

This is inaccurate. Auto-redraw coupling to streams only happens if and only if:

1. You config a stream adapter.
2. You return streams from `render` somewhere in your template.

For your example, if you want to avoid auto-coupling, then simply return the streams' values in your templates and handle the coupling yourself as you're showing.

```js
function View(vm) {
   // manually couple as needed
    stream1.on(vm.redraw.bind(vm))
    stream2.on(vm.redraw.bind(vm))

    return function() {
        return el('input', {disabled: stream1()}, stream2())     // return the *values*
    }
}
```

**Comment by yosbelms** [comment]
^ @leeoniya  Im aware of the solution you mentioned. Im not referring to app-space, I meaning how stremas affect library code. You can see in 86b2e7aac3ff1b1070f52135 how the stream support invades the code in several files.

**Comment by leeoniya** [comment]
If auto-redraw is to be supported (optionally), an `isFunction` check is not enough, because it needs to have `redraw` chained to it after the template is returned. Simply supporting the execution of getters to save you a template-side `()` invocation is not worth the trouble (yes 1.x did this, but i've decided not to support it in 2.x).

There is 0 perf overhead and almost 0 size overhead for supporting BYO streams and it really cleans up app-space code. Again, all of this "magic" is opt-in and provides a lot of benefit for cheap.

**Comment by leeoniya** [comment]
Since streams are just functions, their handling is already supported by normal `on*` event bindings [1]. The one difference would be with parameterized definitions that spread args across the callback: `[myfn, arg1, arg2]`. Streams normally accept a single value, so you'd need to adapt to that if necessary.

[1] https://github.com/leeoniya/domvm/commit/9c39272f3289b7ca26c6cefd4a9b439ae686d042

**Comment by leeoniya** [comment]
`2.x` will no longer provide any ajax/fetch abstractions.

The official recommendation will be to use a minimal promisified xhr lib, like `xr.js` [1] which has a near-identical concise api to 1.x's `w.get`, etc. `xr` minifies to 2.83kb. I've been using it with good success for the threaditjs demo port to 2.x. At first i started with the < 1kb `alite` [2], but ran into issues on failed xhr requests and promise rejections. So 2.x will be BYO ajax (via any xhr/fetch lib), and BYO reactivity (via any stream lib & `domvm.streamCfg`; flyd will be recommended due to its small size).

For the router, I'm still evaluating if domvm's will stick around or should be dumped for an external recommendation like [3]-[7]. I'm particularly interested in small routers that can support both enter & exit (leave) handlers, both history and hash routing, and any that can support href generation (i haven't found one yet like domvm's). Other routers have some nice properties like nested routing, isomorphic support (dunno exactly what this means yet). Depending on their syntax and how much they expose, it may be possible to create an href generator for a third-party routing lib.

[1] https://github.com/radiosilence/xr
[2] https://github.com/chrisdavies/alite
[3] https://visionmedia.github.io/page.js/
[4] https://github.com/flatiron/director
[5] https://github.com/leeluolee/stateman
[6] https://github.com/chrisdavies/rlite
[7] http://microjs.com/#route

**Comment by lawrence-dol** [comment]
I think routing is a pretty central thing to a client app and it's very valuable for DOMVM to provide one. I would not like to see it dropped.

I also think it's a little off-putting to have to pull together half a dozen little projects in order to have a working solution. However, were DOMVM to loosely couple good, but optional, microlibs for these other things that would be an acceptable compromise. Especially with feedback from devs as they find better options. I think that whatever is offered "out of the box" should be enough to build a working client app.

I took a little bit of a look at some of these AJAX microlibs and few seemed to offer much benefit over VanillaJS, especially where Fetch is concerned. So I am not particularly concerned about the loss of the watch module.

**Comment by lawrence-dol** [comment]
Fetch:

  - https://github.com/typicode/fetchival

**Comment by leeoniya** [comment]
> I think routing is a pretty central thing to a client app and it's very valuable for DOMVM to provide one.

Yes, a router is important, but it's also important that if domvm has its own that it provides something tangible over the half-dozen [high quality] others which exist, are well tested, and i don't have to maintain. Since domvm was/is decoupled, its router could always be swapped for another. So a big question is, is there enough that it provides over others to justify maintaining it (assuming it will remain decoupled). The watch module is fully replaced (and improved upon, imo) by streams & an external ajax lib.

> I would not like to see it dropped. I also think it's a little off-putting to have to pull together half a dozen little projects in order to have a working solution. 

It's not half a dozen, it is exactly 2 currently: `flyd`, `xr`. And it would be exactly 3 if a suitable, well-maintained, well-documented router replacement is found which follows domvm's minimalism. I intend to include versions these libs in the domvm repo, so you won't have to hunt for them, similar to how 1.x includes polyfills & a polyfill bundle in the repo.

I know Leo [of Mithril] prefers to write everything himself, including router, neutered promise polyfills, now neutered streams and even his own bundler and testing framework (ospec). But i do not agree with this when well-tested solutions exist. This is especially true for decoupled libs (which Mithril is not).

> https://github.com/typicode/fetchival

I'd like to avoid using fetch as i've mentioned [1], since most of its benefits are for streaming, and it has some drawbacks for more common things like tracking file upload progress & request cancellation.

[1] https://github.com/leeoniya/domvm/issues/93#issuecomment-234311350

**Comment by leeoniya** [comment]
Updates:

- Since domvm is primary focused on `view` aspects, I've flattened the namespace from `domvm.view.*` to `domvm.*`.
- `1.x` router has been imported as-is for now and exposed as `domvm.createRouter()`.
- Several different builds are now available in `/dist`, but it should be easy to pick what you need by referring to [1], and make custom builds if needed (steps are simple and will be posted soon).
- The majority of demos in the repo use either `nano` or `micro` builds, thus my reluctance for having a single monolithic build that incurs up to an extra 53% in added weight over `nano`.

`jsonml` still needs more testing: the plan is to try to run the full test suite from 1.x mostly unchanged by dropping in `2.x` + `jsonml`. Though it won't be a 100% compatibility layer (it wont support getters).

With enough demos ported successfuly and all tests passing, I now consider the view layer to be of beta quality.

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/dist/README.md

**Comment by leeoniya** [comment]
Turns out `2.x` will in fact work with IE10 & IE9 (unofficially at least). There's one failing test that sets `readonly` as a property rather than attribute and IE9 expects the attribute to be a camelcase `readOnly`, so it strangely sets a `readOnly` attribute.

Probably other similar HTML5 prop/attr inconsistencies, but meh.

**Comment by leeoniya** [comment]
added a basic reactive `domvm.prop(val, cb)` implementation to `mini`+ builds. [1]

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/src/utils.js#L120-L129

**Comment by leeoniya** [comment]
i made an experimental branch [1] that adds a hyperscript+spread syntax that allows for children to be *either* explicit arrays or `..restArgs`. It's included in `mini`+ builds and the performance is squarely between regular `domvm.defineElement` and `domvm.jsonml`.

dbmonster @ 0%
```
baseline: 1.23ms
spread:   1.75ms  <--
jsonml:   2.15ms
```

I don't know if it's worth keeping as part of any bundles. It can always live in the repo as an example of something that can be handled in app-space.

[1] https://github.com/leeoniya/domvm/tree/defineElementSpreader

**Comment by leeoniya** [comment]
optimized impl via `domvm.FAST_REMOVE | domvm.FIXED_BODY` flags gets down to 1.09x relative to baseline [1].

https://cdn.rawgit.com/krausest/js-framework-benchmark/4abc882b98aab67d0c7326cc9a7835488a847991/webdriver-ts/table.html

**Comment by lawrence-dol** [comment]
Is this branch "domvm-2" in that table? 

On a side note, I am curious as to why using spread args is any different from using an array arg; are not spreads just syntactic sugar for an array argument?

**Comment by leeoniya** [comment]
> Is this branch "domvm-2" in that table? 

The spread branch? No.

`v2.x-dev` is there as `domvm v2.0.0-beta`

> On a side note, I am curious as to why using spread args is any different from using an array arg; are not spreads just syntactic sugar for an array argument?

Spread args requires more signature testing plus an `Array.prototype.slice.call(arguments, bodyIdx)` after the slicing position is disambiguated. This must be done for *every single node*, so the overhead is quite significant. It's also somewhat fragile since `el("div", "foo")` works fine but `el("div", "foo", "bar")` must be written as `el("div", tx("foo"), tx("bar"))`

**Comment by leeoniya** [comment]
@lawrence-dol there could be a potential path to fragment views via https://github.com/trueadm/fragment-node/issues/1

**Comment by leeoniya** [comment]
I've landed some initial work for `defineFragment` (unrelated to above comment). Right now only initial rendering works; any form of `redraw()` will result in bad things since the dom reconciliation does not handle fragments yet.

The only caveat-free way to implement this was to wrap comment nodes around the fragment. This way there will be no bugs in nesting or adjacent fragments, see demo [1], code [2]. The constructed dom looks like this:

```html
<table>
    <!--1-->
    <thead>
        <tr>
            <th>sku</th>
            <th>price</th>
            <th>qty</th>
        </tr>
    </thead>
    <tbody>
        <!--2-->
        <tr>
            <td>a</td>
            <td>10</td>
            <td>2</td>
        </tr>
        <!--/2-->
        <!--3-->
        <tr>
            <td>c</td>
            <td>30</td>
            <td>1</td>
        </tr>
        <tr class="warn">
            <td colspan="3">^ One Left! ^</td>
        </tr>
        <!--/3-->
        <!--4-->
        <tr>
            <td>b</td>
            <td>20</td>
            <td>3</td>
        </tr>
        <!--/4-->
    </tbody>
    <tfoot></tfoot>
    <!--/1-->
</table>
```

[1] https://cdn.rawgit.com/leeoniya/domvm/2.x-dev/demos/fragments.html
[2] https://github.com/leeoniya/domvm/blob/2.x-dev/demos/fragments.html

**Comment by lawrence-dol** [comment]
Does it make sense for the comments only to exist if there are multiple nodes returned by the view? In the above, only `<!--3-->` is necessary. That could really reduce the DOM size if there were hundreds of rows and only relatively few were a group returned by a single component.

**Comment by leeoniya** [comment]
if a view returns a fragment node then it can contain one or multiple children for which the bounds must be indicated uniformly because the reconciliation algo is very perf sensitive having it handle things in a non-unoform fashion should actually make things slower. the fact that the fragments consist of a single child does not change the internal semantics.

iterating the dom is cheap. and non-rendered nodes are cheap since they dont affect layout/reflow. it just looks big in devtools where it needs to be rendered for you. now with SSR, it would certainly increase the weight of an html string.

like many convenience features (array flattening, jsonml, etc), fragments should be used when necessary and simplify life for you. they're not meant to be free replacements for less burdened 1:1 vtree:vdom structures.

once i have a more functional impl, you should test it out. the overhead should be quite minimal but we'll see.

**Comment by lawrence-dol** [comment]
So, if I understand correctly, if the view must always return either a singular node or a fragment node, presumably using `defineFragment`. It can't sometimes use `defineFragment` and other times not.

**Comment by leeoniya** [comment]
right now yes. in the future maybe not. there have to be some restrictions to what nodes can be recycled one to the other. even root node replacement has a bunch of special handling. now adding to that fragment and root replacement logic makes for very fragile code that's hard to reason about. even in node attr recycling there are limits to how different on* handler forms can be recycled. handling every grafting variation is simply not practical and has shitty perf implications.

**Comment by lawrence-dol** [comment]
Right. I wouldn't work this one too hard. There's a good argument to be made that I should restructure my list component to incorporate the concept of grouping and massage my data to reflect that.

Bear in mind also, that that code is actually written currently for Mithril, so I have some work to do to convert that project to DOMVM when the time comes (thought it definitely will be converted unless Mithril V2 pulls a serious rabbit out of the hat, since I greatly prefer DOMVM).

**Comment by leeoniya** [comment]
I've reverted `defineFragment` changes and split it into its own branch. This rabbit hole ended up being too deep; the simple case is simple but when you have to start accounting for deep fragment-in-fragment paths, everything becomes a searching iterator call to locate the actual parent root and must account for these multiple vnode levels that have no dom representation. Anything that has to do with node re-linking and reordering now has to account for moving blocks of elements, ensuring placeholder nodes are synced during view reuse, etc, etc.

In the interest of actually getting v2 out the door, I'm gonna get back to coding up the good3g email client to iron out the 2.x router semantics.

**Comment by leeoniya** [comment]
Here's where the current 0% overhead registers on Chrome 54:

```
domvm 2.x-dev
-------------
0.10ms (hyperscript w/vm.diff() & FIXED_BODY flag)
0.92ms (hyperscript w/FIXED_BODY flag)
1.17ms (hyperscript)
1.26ms (hyperscript w/array flattening)
2.04ms (jsonml)
2.23ms (jsonml w/array flattening)

domvm 1.x-dev
-------------
3.80ms (jsonml) jsonml
3.93ms (jsonml w/array flattening)

mithril 1.0 (rewrite as of latest commit: 8f1a69c)
---------------------
2.87ms (hyperscript)
2.90ms (hyperscript w/array flattening)
```

One thing that stands out is that Mithril does near-free array flattening. Something that'll be worth looking into.

**Comment by leeoniya** [comment]
I was able to extract `autoPx` into an addon. This is the feature that auto-appends `px` to certain integer `style` props. It saves about 0.4k in weight by excluding the lookup dictionary [1] and reduces the overhead for style-heavy nodes.

This doesn't change the `nano` build (which i intend to keep as the minimum practical build) but this reduced build will form the new `pico` base which itself will be a moving target as things can be split off from nano.

Another thing that may be split from `pico` is parseTag [2] which handles css-style tags like `input.qty[type=number]`.

The dbmonster bench uses no `style` objects, so i switched it and some others over to `pico`.

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/src/view/addons/autoPx.js
[2] https://github.com/leeoniya/domvm/blob/2.x-dev/src/view/addons/parseTag.js

**Comment by leeoniya** [comment]
`diff` now supports returned objects in addition to arrays.

Also, there's now a new optimization opportunity that extends the functionality of `diff`. Previously, `diff` would take a function that returns an array of values to compare between redraws and if those values were the same, then `render()` would be bypassed. A very common situation is when an application has multiple views where only one is visible/active at a time, eg: tabs. This means either you destroy/unmount the current one to load the new one (expensive) or set some `inactive` class for css hide it (much cheaper, retains dom state). But to set `{style: "display: none;"}` or add an `inactive` class would require a full `render()` call and DOM re-sync simply to patch the top-level node. Now there is an opportunity to short-circuit `render()` but still patch the top-level node to hide it and allow `diff` to take care of any followup bypasses. The way it works shown in [1].

I moved some things around based on the premise that `nano` should have access to any needed optimizations. I pulled `vnode.patch` into nano from `micro` and `vm.diff` out of `pico`. The reult is that `micro` only brings the additional `on/emit` to the table, `pico` got smaller and `nano` got fatter.

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/demos/diff-patch.html

**Comment by thysultan** [comment]
@leeoniya how does the `0.10ms (hyperscript w/vm.diff() & FIXED_BODY flag)` impl look like?

**Comment by leeoniya** [comment]
@thysultan 

`FIXED_BODY` is at [1], but it's actually has no effect at 0% mutations cause all subviews' redraws are already short-circuited by the `diff` [2] result. When mutations are non-zero, then it helps with reconciliation since the dom doesnt need a re-ordering/cleanup pass in a fixed structure.

dio looks great btw :)

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/demos/bench/dbmonster/app.js#L9
[2] https://github.com/leeoniya/domvm/blob/2.x-dev/demos/bench/dbmonster/app.js#L67-L113

**Comment by thysultan** [comment]
Q: what VNode will domvm generate for the following `el("span", 100)` or el("span", [1, 2]).

I may be wrong but it looks like you have element shapes though initElementNode just assign node.body without generating the appropriate shape?

**Comment by leeoniya** [comment]
The preprocessor doesn't handle plain text elements in body arrays, only as a full textContent in `el("span", 100)`. You'd have to do `el("span", [tx(1),tx(2)])` where `tx` is `domvm.defineText()`.

**Comment by thysultan** [comment]
Oh that makes sense, though that will mean you forgo patching individual textNodes i.e if tx(1) changes to tx(9) the whole textNode of tx(1) + tx(2) is updated.

btw in the case of dio there is a difference performance wise when using hyperscript like...

```
var tx = dio.VText;
var el = dio.VElement;

el("span", [tx(1), tx(2)])
```

vs

```
h('span', 1, 2)
```

#1 is faster, though the second one is much easier to write and plays well with JSX compilers.

**Comment by leeoniya** [comment]
> though that will mean you forgo patching individual textNodes

right, the preprocessor actually merges adjacent any text nodes anyways (cause they can't be represented in an html string (for SSR) without performing additional [and pointless] comment-insertion gymnastics.).

I'm considering adding a spread signature to `defineElement()`, but that would mean adding plain value handling in the preProc step in addition to more expensive signature disambiguation & argument slicing.body replacement for every single element node. I agree that the the ergonomics are better with spread but the perf tradeoff is non-trivial. The jsonml wrapper in 2.x (which mimics 1.x's template dsl) handles this spread.

Part of the reason for https://github.com/leeoniya/domvm/commit/5c8fe98f1dd52eb6b3c4166340aa38d879539205 (4 days ago) was exactly to explore the spread variation overhead by splitting up `defineElement` and maybe making a `defineElementSpread` with worse perf but better ergonomics.

**Comment by thysultan** [comment]
>the preprocessor actually merges adjacent text nodes anyways (cause they can't be represented in an html string (for SSR) without performing additional [and pointless] comment-insertion gymnastics.).

I also came across this problem, this is how i solved it in the hydration step [1]

[1][https://github.com/thysultan/dio.js/blob/master/src/hydrate.js#L52](https://github.com/thysultan/dio.js/blob/master/src/hydrate.js#L52)

**Comment by leeoniya** [comment]
clever, but i think this is unnecessary complexity (unless your lib's internals have a need for this). in my experience adjacent text nodes are:

1. rare
2. mistakes
3. lazy authors who use `,` rather than `+` :)

just merging them in tpl preproc using a look-behind made the most sense in terms of bytes and simplicity [1]

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/src/view/preProc.js#L56-L63

**Comment by thysultan** [comment]
I could have multiple textNodes mixed with element nodes in an element for example

```
h("span", ["x", h("sub", "2"), "+4x", "+4", "=", "0"])
```

Then trying to change the equation so that "0" becomes "21" without affecting sub, +4x and +4

When it is single textNode i also found that updating .nodeValue vs .textContent or .firstChild.nodeValue is more performant.

**Comment by leeoniya** [comment]
in domvm, once the tpls are preprocessed the internal representation before/after would be

```js
h("span", ["x", h("sub", "2"), "+4x+4=0"])
```

```js
h("span", ["x", h("sub", "2"), "+4x+4=21"])
```

and yes, the actual dom ops performed by domvm would be a single `.nodeValue = "+4x+4=21"`.

**Comment by leeoniya** [comment]
I put a bit of effort into fighting v8 and landed both:

1. plain value element support in body arrays (the tpl preprocessor will now turn these into `domvm.defineText(val)` for you.
2. `domvm.defineElementSpread()` which is able to handle `...restArgs` bodies.

Once I convinced v8 not to de-opt my argument slicing, `defineElementSpread` performance is actually not terrible: `1.43ms` for an impl that includes flattening. Putting is smack between `hyperscript w/flattening` & `jsonml (no flattening)`. I've updated the perf table in the earlier comment.

**Comment by thysultan** [comment]
```
body = sliceArgs(args, bodyIdx);
```
v8 could be smart enough to know you are passing it by reference, you might get better results if you build the array by hand.

**Comment by grumpi** [comment]
Good work!

It looks like there is a small issue with `vm.html()` when not rendering to DOM first. It appears that all the subviews get rendered as `</>` when directly using `vm.html()` (as you would do in a server-side rendering setting).

Simplest way to reproduce: change one of the subview tests in the test suite (I tried "Subview List w/keys - Create") as follows:

1. don't mount.
2. `assert.equal(vm.html(), expcHtml, "vm.html(): "+ vm.html());`

----

Apart from that, I think I've successfully ported my code to domvm 2.x-dev.

Things I had to do:

1. `.route()` -> `.createRouter()`.
2. `.view()` -> `.createView()`.
3. replace all `domvm.watch().prop` by `domvm.prop`. Very cool to have it decoupled.
4. convert templates in the obvious way. I like the hyperscript templates better since it's much clearer what's going on.
5. replace `w.fetch()` with something else.
6. attributes like `{value: property}` needed to be changed to `{value: property()}` since domvm doesn't seem to evaluate functions in attributes anymore. This is good.
7. explicitly set onclick attribute on links to do `e.preventDefault(); router.goto()` to get to the target location without a page reload. At first I was confused why things didn't work, but after looking into the code and realizing what changed, I like this new way of explicitly defining how navigation works.
8. add `global.document = null;` on the server-side code (nodejs) in order to avoid triggering an error on the line `var doc = document;` (in domvm). On the positive side, I was able to remove the line `global.history = {};` from my code since the new domvm runs fine without that one.

I think that was all that was needed to get it to run again. I suppose there's a few things I should sort out in terms of view redrawing, i.e. using the new `vm.parent().redraw()`, and such.

**Comment by leeoniya** [comment]
@thysultan 

> you might get better results if you build the array by hand.

`sliceArgs()` *does* build the array by hand. Before it would use `Array.prototype.slice.call`, which caused the de-opt. I also profiled pulling the code out of the function, using a fixed length array, but V8 is smart enough to inline everything it needs at this point.

@grumpi 

> It looks like there is a small issue with vm.html() when not rendering to DOM first.

thanks, will look into this.

> replace `w.fetch()` with something else.

yep, [xr](https://github.com/radiosilence/xr) is my recommendation

> domvm doesn't seem to evaluate functions in attributes anymore

correct, it saves on perf and internally type-checking everything everywhere. explicit invocation is clearer on the template side also.

> `.route()` -> `.createRouter()`....
> explicitly set onclick attribute on links to do e.preventDefault(); router.goto() to get to the target location without a page reload. 

the router and routing semantics are still being refined from 1.x, so some of this may not be intentional and will be subject to change.

> add global.document = null; on the server-side code (nodejs) in order to avoid triggering an error

thanks, will fix.

Glad to hear the transition was mostly smooth and positive.

On a side note, i'll probably move `jsonml` out of the `full` build and just make it a stand-alone js file. Now that there's `defineElementSpread` and plain value support in array bodies, its use case is really niche.

**Comment by grumpi** [comment]
@leeoniya xr sounds super cool since it's so lightweight. I'm currently using https://github.com/mzabriskie/axios (which is a lot heavier), but only because it has a uniform API (including timeout) across browser and nodejs.

Going to look again into finding a reasonable nodejs implementation of XMLHttpRequest. That, together with xr sounds like a better alternative when server-side rendering is needed.

**Comment by leeoniya** [comment]
the SSR bugs should be fixed.

**Comment by leeoniya** [comment]
...continued in #104.
